domain/modelmigration/service/service.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package service
+
+import (
+	"context"
+
+	"github.com/juju/version/v2"
+
+	coreerrors "github.com/juju/juju/core/errors"
+	"github.com/juju/juju/core/providertracker"
+	"github.com/juju/juju/domain/modelmigration"
+	"github.com/juju/juju/environs/envcontext"
+	environscontext "github.com/juju/juju/environs/envcontext"
+	"github.com/juju/juju/environs/instances"
+	"github.com/juju/juju/internal/errors"
+)
+
+// InstanceProvider describes the interface that is needed from the cloud provider to
+// implement the model migration service.
+type InstanceProvider interface {
+	AllInstances(envcontext.ProviderCallContext) ([]instances.Instance, error)
+}
+
+// ResourceProvider describes a provider for managing cloud resources on behalf
+// of a model.
+type ResourceProvider interface {
+	// AdoptResources is called when the model is moved from one
+	// controller to another using model migration. Some providers tag
+	// instances, disks, and cloud storage with the controller UUID to
+	// aid in clean destruction. This method will be called on the
+	// environ for the target controller so it can update the
+	// controller tags for all of those things. For providers that do
+	// not track the controller UUID, a simple method returning nil
+	// will suffice. The version number of the source controller is
+	// provided for backwards compatibility - if the technique used to
+	// tag items changes, the version number can be used to decide how
+	// to remove the old tags correctly.
+	AdoptResources(envcontext.ProviderCallContext, string, version.Number) error
+}
+
+// Service provides the means for supporting model migration actions between
+// controllers and answering questions about the underlying model(s) that are
+// being migrated.
+type Service struct {
+	// instanceProviderGetter is a getter for getting access to the models
+	// [InstanceProvider].
+	instanceProviderGetter func(context.Context) (InstanceProvider, error)
+
+	// resourceProviderGetter is a getter for getting access to the models
+	// [ResourceProvider]
+	resourceProviderGettter func(context.Context) (ResourceProvider, error)
+
+	st State
+}
+
+// State defines the interface required for accessing the underlying state of
+// the model during migration.
+type State interface {
+	GetControllerUUID(context.Context) (string, error)
+}
+
+// New is responsible for constructing a new [Service] to handle model migration
+// tasks.
+func New(
```

Comment:
Every other domain service uses `NewService`.
There's also `NewWatchableService` everywhere.
Would prefer consistency either way. If we adopt `New`, what does `NewWatchableService` become.

Reply:
That is fine.

Reply:
Done & Done